_meeting-notes/2019-10-22.md

```diff
@@ -0,0 +1,128 @@
+Test Ground Weekly
+===
+
+###### tags: `IPFS` `Testing`
+
+:::info
+- **Host:** @daviddias
+- **Location:** https://protocol.zoom.us/my/daviddias
+- **Date:** Oct 22, 2019 5:00 PM (UTC)
+- **Participants:**
+    - @daviddias
+    - @jimpick
+    - @stebalien
+    - @momack2
+- **Agenda**
+  1. Sprint Update `20min`
+  2. Identify blockers / Answer Questions `20min`
+  3. This week's planning `20min`
+- **Reference:** 
+  - [Last week meeting minute](https://github.com/ipfs/testground/pull/69)
+  - [Trello Board](https://trello.com/b/O9z3ljaH/testground-ignition-%F0%9F%9A%80)
+  - [Testing Infra Design](https://github.com/ipfs/testground/blob/master/docs/SPEC.md)
+  - [Testing Inventory](https://github.com/ipfs/testground/blob/master/docs/test-inventory.md)
+  - [Testing Improvement Tracker](https://docs.google.com/spreadsheets/d/1xyqyGUF-oe3x9ln88YonVeOMWWdknik74lVgL_3dBY8/edit#gid=0)
+
+:::
+
+## :mega: Update (Done, Shipped, Made Progress)
+
+- ==@daviddias==
+  - Completed the "go-ipfs as a Library" Tutorial, now in review phase by the docs Team https://github.com/ipfs/go-ipfs/pull/6695/
+  - Landed smlbench working with TestGround orchestrator (thank you Jim for helping with the last mile there ❤️) https://trello.com/c/1LTXALLW/62-get-smlbench-plan-to-work-with-testground-orchestrator
+  - Continue iterating on the Test Plan spec  https://trello.com/c/0erPZId8/43-specify-10-test-plans
+    - now with differentiation for Plan & Test Parameters https://github.com/ipfs/testground/pull/76
+    - now each Test Plan lives with the plan folder of the Plan itself https://github.com/ipfs/testground/pull/77
+  - Revisited the OKRs as requested https://trello.com/c/o09gs18G/63-review-and-iterate-on-q4-okrs
+  - Created a draft Roadmap for review https://github.com/ipfs/testground/pulls
+  - Got sick, then got better :) ETOOMUCHLONGFLIGHTS
+  - Broke a tooth 🦷 (tip: watch out for those Swiss Ricolas!). **FYI** will be traveling this week to Lisbon for a emergency dentist appt (and a reschedule of others I was going to reschedule to post hack week, but this way I merge it all together rather than going to Lisbon 2x). This means that I'm almost at 2 days of lost work given flying + transportation + being at the dentist.
+- ==@jimpick==
+  - smlbench (adds a randomly generated file on one node, gets it from a 2nd node)
+    - initial work, one docker container, IPTB with 2 IPFS nodes https://github.com/ipfs/testground/pull/74
+    - follow-on work, two dockers containers, IPTB with 1 IPFS node per container, using Sync SDK (redis) to co-ordinate https://github.com/ipfs/testground/pull/74
+  - Lots of little puzzles solved:
+    - IPFS into Docker using Docker.template and conditional via shell script (works, but not satisfied)
+    - IPTB SDK usage
+    - Sync SDK usage
```

Comment:
What did you learn @jimpick ?

Reply:
Sync SDK is here:

https://godoc.org/github.com/ipfs/testground/sdk/sync

It's the code that communicates between Docker containers using Redis.

The first thing I do with it is register each peer using `writer.Write()` ... it expects a peer.AddrInfo structure with the Peer ID and Multiaddrs - because I'm writing to the sync.PeerSubtree location, that enforces the type check and generates the prefix for the Redis key, eg. "run::plan::case:simple-add-get:nodes:..."

https://github.com/ipfs/testground/blob/f3746e82ce5ed06f93632bcb502d94e216c2ce3a/plans/smlbench2/main.go#L86

When I write to the sync API, it returns a sequence number. I have 2 peers running in separate Docker containers ... if the sequence number returns "1", then I make that peer act as the *"adder"* node. If the sequence is "2", then I make that peer into the *"getter"* node.

# The *"adder"* node goes first

It doesn't have to wait for anything, so it just makes some random data and adds it to the IPFS node.

We need to write the CID of the random data into Redis so the *"getter"* node can figure out what to fetch. We do that with `writer.Write()`:

https://github.com/ipfs/testground/blob/f3746e82ce5ed06f93632bcb502d94e216c2ce3a/plans/smlbench2/main.go#L123

I define a tiny namespace in Redis for it to store the state here:

https://github.com/ipfs/testground/blob/f3746e82ce5ed06f93632bcb502d94e216c2ce3a/plans/smlbench2/main.go#L25-L31

That will generate a key prefix in Redis that looks like `run::plan::case:simple-add-get:cid:...` and enforces the type and does the marshalling/unmarshalling for that automatically between when it is written from one Docker instance and read from another.

Once the CID is written, you can see that the "adder" node signals that it has reached the **"added"** state here:

https://github.com/ipfs/testground/blob/f3746e82ce5ed06f93632bcb502d94e216c2ce3a/plans/smlbench2/main.go#L129

At this point, it can't just exit, because the IPFS node needs to stay running so the *"getter"* node can fetch data from it. So it waits until the *"getter"* node signals that it has reached the **"received"** state using `watcher.Barrier()`:

https://github.com/ipfs/testground/blob/f3746e82ce5ed06f93632bcb502d94e216c2ce3a/plans/smlbench2/main.go#L136

# The *"getter"* node goes next

Once the *"getter"* node knows that it is the 2nd node, it has to get the multiaddr of the *"adder"* node so it can use IPFS to connect to it.

It uses the `watcher.Subscribe()` SDK method to get a channel of updates for the peers that have added their information to Redis. Since the *"adder"* node has already added data (even if it was added before the second node was even started) - the SDK will fetch that from Redis and immediately output an event on the channel with the contents of the peer.Addrinfo data structure ... all nicely unmarshalled and typed! (very nice DX @raulk, a joy to use).

https://github.com/ipfs/testground/blob/f3746e82ce5ed06f93632bcb502d94e216c2ce3a/plans/smlbench2/main.go#L159

With that, it's easy to call the the IPFS API to connect to the peer.

Once connected, the *"getter"* node has to wait for the *"adder"* node to reach the **"added"** state. That's done by setting a "barrier" using `watcher.Barrier()`

https://github.com/ipfs/testground/blob/f3746e82ce5ed06f93632bcb502d94e216c2ce3a/plans/smlbench2/main.go#L188

Once it sees the **"added"** state has been reached, it fetches the CID using `watcher.Subscriber()` again, but this time looking at the Redis namespace set aside for the CID (instead of the AddrInfo space as before):

https://github.com/ipfs/testground/blob/f3746e82ce5ed06f93632bcb502d94e216c2ce3a/plans/smlbench2/main.go#L197

The *"getter"* node now has all the info it needs, so it can just call the IPFS API to fetch the content and time how long it takes.

The last thing the *"getter"* node does is signal that it has reached the **"received"** state.

https://github.com/ipfs/testground/blob/f3746e82ce5ed06f93632bcb502d94e216c2ce3a/plans/smlbench2/main.go#L221

It just exits at this point because it is done.

# The *"adder"* node goes last

The adder node has been waiting for the "received" state using `watcher.Barrier()` ... now that the **"received"** state has been reached, there's nothing more to do, so it can safely exit (shutting down its IPFS node).

Reply:
Hey @jimpick, it would be great to add these explanations on the godocs. They are just not very discoverable here. Mind taking a quick stab at that?